phost/src/main.rs

```diff
@@ -404,29 +431,45 @@ async fn batch_sync_block(
         let r = req_sync_header(pr, &header_batch, authrotiy_change.as_ref()).await?;
         println!("  ..sync_header: {:?}", r);
 
-        let dispatch_window = batch_window - 1;
-        while !block_batch.is_empty() {
-            // TODO: fix the potential overflow here
-            let end_batch = block_batch.len() as isize - 1;
-            let batch_end = cmp::min(dispatch_window as isize, end_batch);
-            if batch_end >= 0 {
-                let mut dispatch_batch: Vec<BlockHeaderWithEvents> = block_batch
-                    .drain(..=(batch_end as usize))
-                    .map(|bwe| BlockHeaderWithEvents {
-                        block_header: bwe.block.block.header,
-                        events: bwe.events,
-                        proof: bwe.proof,
-                        worker_snapshot: None,
-                    })
-                    .collect();
-                maybe_take_worker_snapshot(client, events_decoder, &mut dispatch_batch).await?;
-                let r = req_dispatch_block(pr, &dispatch_batch).await?;
-                println!("  ..dispatch_block: {:?}", r);
-
-                // Update sync state
-                synced_blocks += dispatch_batch.len();
-            }
-        }
+        let para_fin_header = chain_client::get_parachain_heads(&client, &paraclient, Some(last_header_hash)).await;
+		if para_fin_header.is_some() {
+			let para_fin_hash = &para_fin_header.unwrap()[2..34];
```

Comment:
Please check the problem in `get_parachain_heads` first. In the past, `para_fin_header` refers to the raw storage. It is a `Vec<u8>`, and the first 2 bytes encode the size of the vector. That's why you skipped the first 2 bytes.

However, although `Para.Heads` is a `Vec<u8>`. It's actually a block header. The first 32 bytes of a block header is the `parent_hash`. So you extracted the parent hash, not the latest block hash.

To get the latest block hash, simply calculate `blake2_256(para_heads)`. You can find blake2_256 from `sp_core` (don't use sp_io!). Or you can decode it with a block header and call `hash()`

Example:
https://github.com/Phala-Network/phala-blockchain/blob/e7a71a3257622949194d2e1bfa23b136e923cac6/scripts/debug-cli/src/main.rs#L66-L70

Reply:
It worked